desc/objectives/_free_boundary.py

```diff
@@ -754,6 +755,338 @@ def _print(fmt, fmax, fmin, fmean, norm, units):
+        if self._s is None:
+            k = min(src_grid.num_theta, src_grid.num_zeta)
+            self._s = k // 2 + int(np.sqrt(k))
+        if self._q is None:
+            k = min(src_grid.num_theta, src_grid.num_zeta)
+            self._q = k // 2 + int(np.sqrt(k))
+
+        try:
+            interpolator = FFTInterpolator(eval_grid, src_grid, self._s, self._q)
+        except AssertionError as e:
+            warnings.warn(
+                "Could not built fft interpolator, switching to dft method which is"
+                " much slower. Reason: " + str(e)
+            )
+            interpolator = DFTInterpolator(eval_grid, src_grid, self._s, self._q)
+
+        self._constants = {"interpolator": interpolator}
+        self._data_keys = [
+            "R",
+            "zeta",
+            "Z",
+            "n_rho",
+            "K_vc",
+            "phi",
+        ]
+        self._args = get_params(
+            self._data_keys,
+            obj="desc.equilibrium.equilibrium.Equilibrium",
+            has_axis=False,
+        )
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        src_profiles = get_profiles(self._data_keys, obj=eq, grid=src_grid)
+        src_transforms = get_transforms(self._data_keys, obj=eq, grid=src_grid)
+        eval_profiles = get_profiles(self._data_keys, obj=eq, grid=eval_grid)
+        eval_transforms = get_transforms(self._data_keys, obj=eq, grid=eval_grid)
+
+        w = eval_grid.weights
+        w *= jnp.sqrt(eval_grid.num_nodes)
+
+        self._constants.update(
+            eval_transforms=eval_transforms,
+            eval_profiles=eval_profiles,
+            src_transforms=src_transforms,
+            src_profiles=src_profiles,
+            quad_weights=w,
+        )
+
+        if self._vacuum:
+            eval_data = compute_fun(
+                "desc.equilibrium.equilibrium.Equilibrium",
+                self._data_keys,
+                params=eq.params_dict,
+                transforms=eval_transforms,
+                profiles=eval_profiles,
+            )
+            Bplasma = np.zeros(np.shape(eval_grid.nodes))
+            self._constants.update(Bplasma=Bplasma, eval_data=eval_data)
+
+        elif self._eq_fixed:
+            eval_data = compute_fun(
+                "desc.equilibrium.equilibrium.Equilibrium",
+                self._data_keys,
+                params=eq.params_dict,
+                transforms=eval_transforms,
+                profiles=eval_profiles,
+            )
+            src_data = compute_fun(
+                "desc.equilibrium.equilibrium.Equilibrium",
+                self._data_keys,
+                params=eq.params_dict,
+                transforms=src_transforms,
+                profiles=src_profiles,
+            )
+
+            # don't need extra B/2 since we only care about normal component
+            Bplasma = -singular_integral(
```

Comment:
see the BoundaryError objective, I added a convenience wrapper for the biot-savart integral